framework/src/samplers/Sampler.C

```diff
@@ -86,3 +116,38 @@ Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
 
   _generator.saveState();
 }
+
+void
+Sampler::setSampleNames(const std::vector<std::string> & names)
+{
+  _sample_names = names;
+
+  // Use assert because to check the size a getSamples call is required, which you don't
+  // want to do if you don't need it.
+  mooseAssert(getSamples().size() == _sample_names.size(),
+              "The number of sample names must match the number of samples returned.");
+}
+
+Sampler::Location
+Sampler::getLocation(unsigned int global_index)
+{
+  if (_offsets.empty())
+    reinit(getSamples());
+
+  mooseAssert(_offsets.size() > 1,
+              "The getSamples method returned an empty vector, if you are seeing this you have "
+              "done something to bypass another assert in the 'getSamples' method that should "
+              "prevent this message.");
+
+  std::vector<unsigned int>::const_iterator iter =
+      std::upper_bound(_offsets.begin(), _offsets.end(), global_index) - 1;
```

Comment:
std::lower uses not less, which when equal does the wrong thing. std::upper uses greater than, so the equal case can't happen.